### PR TITLE
deprecate provides and requires in nlu

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- deprecate provides and requires in nlu
- use `required_components()` instead

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/components.py [hunk 1]
```diff
@@ -68,89 +67,64 @@ def validate_tokenizers(pipeline: List["Component"]) -> None:
...
+    """Validates that all required components are present in the pipeline."""
 
-def validate_requires_any_of(
-    required_properties: Tuple[Text],
-    provided_properties: Set[Text],
-    component_name: Text,
-) -> None:
-    """Validates that at least one of the given required properties is present."""
+    for i, component in enumerate(pipeline):
+        if hasattr(component, "provides"):
```
Comment (tabergma) [inline rasa/nlu/components.py]: Please move the checks for deprecated attributes to a separate method.

File: rasa/nlu/components.py [hunk 2]
```diff
@@ -68,89 +67,64 @@ def validate_tokenizers(pipeline: List["Component"]) -> None:
...
+                docs="https://rasa.com/docs/rasa/migration-guide/",
+            )
 
-    property_present = any(
-        [property in provided_properties for property in required_properties]
-    )
+        missing_components = []
+        for required_component in component.required_components():
+            if not _required_component_in_pipeline(
+                required_component, pipeline[: i + 1]
```
Comment (tabergma) [inline rasa/nlu/components.py]: Why `i + 1`?

File: rasa/nlu/components.py [hunk 3]
```diff
@@ -68,89 +67,64 @@ def validate_tokenizers(pipeline: List["Component"]) -> None:
...
-
-def validate_arguments(
-    pipeline: List["Component"],
-    context: Dict[Text, Any],
-    allow_empty_pipeline: bool = False,
-) -> None:
-    """Validates that all arguments are present to train the pipeline."""
-
-    # Ensure the pipeline is not empty
-    if not allow_empty_pipeline and len(pipeline) == 0:
```
Comment (tabergma) [inline rasa/nlu/components.py]: Should we keep the check for an empty pipeline?

File: rasa/nlu/extractors/extractor.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+from typing import Any, Dict, List, Text, Tuple
+
+from rasa.nlu.components import Component
+from rasa.nlu.constants import EXTRACTOR, ENTITIES
+from rasa.nlu.training_data import Message
+
+
+class EntityExtractor(Component):
```
Comment (tabergma) [inline rasa/nlu/extractors/extractor.py]: Did you change anything in here? I assume not.

File: rasa/nlu/components.py (hunk 2 above)
Comment (Ghostvv) [reply rasa/nlu/components.py]: good point, I guess a component can't require itself

File: rasa/nlu/components.py (hunk 3 above)
Comment (Ghostvv) [reply rasa/nlu/components.py]: I don't know, why did we add it?

Comment (Ghostvv) [reply rasa/nlu/components.py]: yes, thanks for spotting it

File: rasa/nlu/extractors/extractor.py (hunk 4 above)
Comment (Ghostvv) [reply rasa/nlu/extractors/extractor.py]: no, just copy pasted to a new file

File: changelog/5266.misc.rst [hunk 5]
```diff
@@ -7,3 +7,6 @@ include in your pipeline.
 Use ``DIETClassifier`` instead of ``SklearnIntentClassifier``.
 
 Use ``TEDPolicy`` instead of ``KerasPolicy``.
+
+Properties ``Component.provides`` and ``Component.requires`` are deprecated.
```
Comment (tmbo) [inline changelog/5266.misc.rst]: This shouldn't be misc (as this is relevant for the user) - I think the note should be `5266.removal.rst`

**Pull request merged**